metrotranscom/doorway
# refactor: add asset->url functions and implement in public

# Pull Request Template

## Issue Overview

This PR addresses (but does not fully resolve) issue #78 


## Description

This PR adds two new functions, `getImageUrlFromAsset` and `getPdfUrlFromAsset` to `shared-helpers` and implements them in the public site.  These functions are intended to implement URL resolution at the point where it actually matters (`Asset` objects) and provide a single point for each file type where that logic can be defined.  This also makes it easier to implement uploads on the backend and consolidate URL resolution logic there by defining an initial case in each function that would enable the backend to explicitly define the URL to use.

The amount of changes needed to implement this abstraction layer in the partner site (including additional abstractions around uploads) outnumber the combined changes here, so they will be introduced in a subsequent PR.


## How Can This Be Tested/Reviewed?

Rebuild and start up the public site and navigate to the listings page and some individual listings.  These changes should be fully transparent; if you don't notice anything different, it means they're working.

New unit tests were added to verify the functionality added by these functions.  Unit tests were also ran on the public site to verify changes were non-breaking.  Run `yarn test` in `shared-helpers` and `yarn test:unit` under `sites/public` to confirm.

## Checklist:


## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.
